### PR TITLE
Switch to Python3.

### PR DESCRIPTION
**Pull request by chenshuo**

### CODE REVIEW

**Comment by zuxy** [comment]
Is "/usr/bin/env python3" better than "/usr/bin/python3"?

**Comment by chenshuo** [comment]
> Is "/usr/bin/env python3" better than "/usr/bin/python3"?
Yes. Updated code.

File: gtests/net/tcp/common/set_sysctls.py [hunk 1]
```diff
@@ -1,24 +1,46 @@
...
+__author__ = ('brakmo@google.com (Lawrence Brakmo)')
 
-import commands
 import os
+import subprocess
 import sys
 
-pppid = int(os.popen("ps -p %d -oppid=" % os.getppid()).read().strip())
-filename = '/tmp/sysctl_restore_%d.sh' % pppid
+# Sometimes shell skips fork() as an optimization.
```
Comment (nealcardwell) [inline gtests/net/tcp/common/set_sysctls.py:19]: This part is different from internal packetdrill. It would be good to keep this part of packetdrill upstream and internally in sync. Is the plan to backport this commit to our internal packetdrill after it is merged upstream? ( If so, this sounds fine to me. :-) )

File: gtests/net/tcp/common/set_sysctls.py [hunk 2]
```diff
@@ -1,24 +1,46 @@
...
+# https://unix.stackexchange.com/questions/466496/
+# If your shell does so, define SHELL_SKIPS_FORK=1
```
Comment (nealcardwell) [inline gtests/net/tcp/common/set_sysctls.py:21]: Interesting! For my education, on which OS distribution and shell did you run into this issue?

File: gtests/net/tcp/common/set_sysctls.py [hunk 3]
```diff
@@ -1,24 +1,46 @@
...
+# before running packetdrill.
+if os.environ.get('SHELL_SKIPS_FORK'):
+  filename = '/tmp/sysctl_restore_%d.sh' % os.getppid()
+else:
+  pppid = int(os.popen("ps -p %d -oppid=" % os.getppid()).read().strip())
+  filename = '/tmp/sysctl_restore_%d.sh' % pppid
```
Comment (nealcardwell) [inline gtests/net/tcp/common/set_sysctls.py:27]: To reduce code duplication can we use something like:

    if os.environ.get('SHELL_SKIPS_FORK'):
        id = os.getppid()
    else:
        id = int(os.popen("ps -p %d -oppid=" % 
                      os.getppid()).read().strip())
    filename = '/tmp/sysctl_restore_%d.sh' % id

**Review by nealcardwell — COMMENTED** [review]
Thanks, Shuo!

Style nit: for consistency, the first line of the commit, can you please use something like:

net-test: packetdrill: switch to python3

Thanks!

File: gtests/net/tcp/common/set_sysctls.py (hunk 2 above)
Comment (chenshuo) [reply gtests/net/tcp/common/set_sysctls.py:21]: On Debian 11 and Ubuntu, /bin/sh points to `dash`, which doesn't skip fork(), so I get process tree like:
```
virtme-init─┬─bash───packetdrill─┬─sh───python3
            │                    │ ^^ Here's /bin/sh by system(3)
            │                    └─{packetdrill}
            └─systemd-udevd───2*[systemd-udevd]
```
In this case, `os.getppid()` returns pid of `sh`, not `packetdrill`, so one additional step is needed to find **pppid** (line 26 below). This is the original behavior of this script (the open source version). I believe that's the reason Willem introduced **pppid** in commit 51a1a06b.

However, in Google, `/bin/sh` points to `bash`, which **does** skip fork() for simple command like this. So `os.getppid()` returns pid of `packetdrill`:

```
virtme-init─┬─bash───packetdrill─┬─python3
            │                    └─{packetdrill}
            └─systemd-udevd───2*[systemd-udevd]
```
So the internal version of this script uses `os.getppid()` alone, no need to call `ps` command.

I didn't know any programmtic way to detect shell's behavior, or disable this optimization in `bash`.
I introduced `SHELL_SKIPS_FORK` so that I could test inside Google before uploading to github.com.
This is far from ideal, and I have no idea how to keep this part of packetdrill upstream and internally in sync :-(

Suggestions?

Comment (nealcardwell) [reply gtests/net/tcp/common/set_sysctls.py:21]: Thanks for the detailed explanation! That's very clear and very helpful.

It seems like it will not be a good user experience if packetdrill users have to know whether their shell skips fork, and thus know whether to set the SHELL_SKIPS_FORK environment variable. :-)

I would suggest that we change the set_sysctls.py approach so that it does not rely on PPID. I think Larry Brakmo just used that for the set/restore scheme because it was easy. But now that commonly used shells have diverged in this way, it's no longer easy, so I think we should just change approaches. :-)

Here's a proposal: I'd suggest a small patch to packetdrill to export a  unique ID for the test into an environment variable, and set_sysctls.py and the restore command can both use that to coordinate. We could use many different approaches to generate the testid, but since we know the pid is simple and works, I guess we can start with that. So how about something like: 

1: packetdrill can run something like:
 (a) the following in state_new() in run.c:
  asprintf(&state->test_id, "%d", getpid());
  setenv("PACKETDRILL_TEST_ID", state->test_id, 1);
 (b) the following in state_free():
  free(state->test_id);
 (c) test_id can be defined in the state struct in run.h with something like:
   char *test_id; /* opaque id string for current test */

2: set_sysctls.py can run:
  filename = '/tmp/sysctl_restore_%s.sh' % os.getenv('PACKETDRILL_TEST_ID')

3: for scripts that restore the sysctl values, we can change:
    `/tmp/sysctl_restore_${PPID}.sh`
  to:
    `/tmp/sysctl_restore_${PACKETDRILL_TEST_ID}.sh`

  You should be able to change all of the existing sysctl_restore_ invocations with something like:
` find gtests/net/ -name '*.pkt' |  xargs perl -i.orig -p -e 's|/tmp/sysctl_restore_\${PPID}.sh|/tmp/sysctl_restore_\${PACKETDRILL_TEST_ID}.sh|g'
`
 
I would suggest testing this briefly on one test upstream and then one test in 9xx. If it works on one test upstream and 9xx, then I'd suggest testing it for all //net/tcp/... and //net/gre/... and //net/packetdrill/... tests in 9xx, and then sending it for review in 9xx. Then we can send it upstream. Thanks!

@wdebruij  for visibility and in case Willem has other ideas.

File: gtests/net/tcp/common/set_sysctls.py (hunk 3 above)
Comment (nealcardwell) [reply gtests/net/tcp/common/set_sysctls.py:27]: With the PACKETDRILL_TEST_ID proposal above, we can ignore this part of the review. :-)

File: gtests/net/tcp/common/set_sysctls.py (hunk 2 above)
Comment (chenshuo) [reply gtests/net/tcp/common/set_sysctls.py]: Thanks for the suggestion! I updated this pull request:
1. call `setenv(name, value, 1)` in main(). According to `man 3 setenv` makes copies of the strings pointed to by name and value (by contrast with `putenv(3)`), so no need to add a `test_id` field to struct `state`.
2. updated `set_sysctls.py` as you described.
3. no need to update `.pkt` files as long as `PACKETDRILL_TEST_ID` is set to pid of packetdrill. Because the substitution of `${PPID}` happens inside the `sh` started by `system(3)`, its `PPID` points to packetdrill no matter what shell we are using.

I tested this approach works for both `bash` and `dash` on my own machine. I will set it for review in 9xx too.

File: gtests/net/tcp/common/set_sysctls.py (hunk 1 above)
Comment (chenshuo) [reply gtests/net/tcp/common/set_sysctls.py]: Now this file is in sync with internel version :-)

File: gtests/net/tcp/common/set_sysctls.py (hunk 3 above)
Comment (chenshuo) [reply gtests/net/tcp/common/set_sysctls.py]: Ack.

File: gtests/net/packetdrill/packetdrill.c [hunk 4]
```diff
@@ -61,12 +61,25 @@ static void run_init_scripts(struct config *config)
 	free(scripts);
 }
 
+/* Sets "PACKETDRILL_PID" environment variable to pid of this process,
```
Comment (nealcardwell) [inline gtests/net/packetdrill/packetdrill.c:64]: The PACKETDRILL_PID changes in here seem to be unrelated to the stated subject of the commit. In the Linux git ecosystem we try to keep commits as small and narrowly-focused as possible, and we try to accomplish things, where possible, using clean cherry-picks of existing commits elsewhere.

Please pull out the PACKETDRILL_PID diffs with something like the following:

git checkout packetdrill/master
git cherry-pick <internal_PACKETDRILL_PID_sha1>
 < clean up internal footers, leaving Change-Id, and add signed-off-by >
git cherry-pick 733f92fa5ddd25aea76d2d6d655277dfff0cfc93

File: gtests/net/packetdrill/packetdrill.config [hunk 5]
```diff
@@ -3,3 +3,5 @@
 # ./scripts/kconfig/merge_config.sh -m .config packetdrill.config
 
 CONFIG_TUN=y
+CONFIG_NET_SCH_FQ=y
```
Comment (nealcardwell) [inline gtests/net/packetdrill/packetdrill.config:6]: these two lines seem to be unrelated to python3... what's the motivation here? can they be a cherry-pick from some other commit? if we need them, they should be a separate commit.

**Review by nealcardwell — COMMENTED** [review]
thanks!

File: gtests/net/tcp/common/set_sysctls.py [hunk 6]
```diff
@@ -1,24 +1,42 @@
...
+  # read current value and add restore command to file
+  cur_val = subprocess.check_output(['cat', sysctl[0]], universal_newlines=True)
+  print('echo "%s" > %s' % (cur_val.strip(), sysctl[0]), file=restore_file)
 
   # set new value
   cmd = 'echo "%s" > %s' % (sysctl[1], sysctl[0])
   os.system(cmd)
 
 os.system('chmod u+x %s' % filename)
+
```
[the remark this answers is not in the document]
Comment (chenshuo) [reply gtests/net/tcp/common/set_sysctls.py]: done.

File: gtests/net/packetdrill/packetdrill.c (hunk 4 above)
Comment (chenshuo) [reply gtests/net/packetdrill/packetdrill.c:64]: done. split into two commtis.

File: gtests/net/packetdrill/packetdrill.config (hunk 5 above)
Comment (chenshuo) [reply gtests/net/packetdrill/packetdrill.config]: I removed this change from this patch set.
It will be needed when upload more test cases.

File: gtests/net/packetdrill/config.c [hunk 7]
```diff
@@ -221,7 +221,7 @@ static void set_ipv6_defaults(struct config *config)
 void set_default_config(struct config *config)
 {
 	memset(config, 0, sizeof(*config));
-	config->code_command_line	= "/usr/bin/python";
+	config->code_command_line	= "python3";
```
Comment (zuxy) [inline gtests/net/packetdrill/config.c:224]: any particular reason why we remove /usr/bin?

**Review by nealcardwell — COMMENTED** [review]
Thank you, Shuo!

File: gtests/net/tcp/common/set_sysctls.py [hunk 8]
```diff
@@ -1,21 +1,38 @@
-#!/usr/bin/python
+#!/usr/bin/env python3
```
[the remark this answers is not in the document]
Comment (chenshuo) [reply gtests/net/tcp/common/set_sysctls.py:1]: Done.
The 1st commit was to update set_sysctls.py to internal SHA1 0c991040df7c5, the one right before switching to python3.
The 3rd commit switches scripts to Python3.

**Comment by nealcardwell** [comment]
Looks good. Thanks for doing this!
